securesystemslib/metadata.py

```diff
@@ -97,3 +101,87 @@ def pae(self) -> bytes:
             len(self.payload),
             self.payload,
         )
+
+    def sign(self, signer: Signer) -> Signature:
+        """Sign the payload and create the signature.
+
+        Arguments:
+            signer: A "Signer" class instance.
+
+        Raises:
+            TypeError: If "signer" is not an instance of the "Signer" class.
+
+        Returns:
+            A "Signature" instance.
+        """
+
+        if not isinstance(signer, Signer):
+            raise TypeError(f"expected type 'Signer', got {type(signer).__name__}")
+
+        signature = signer.sign(self.pae)
+        self.signatures.append(signature)
+
+        return signature
+
+    def verify(self, keys: KeyList, threshold: int) -> List[str]:
+        """Verify the payload with the provided Keys.
+
+        Arguments:
+            keys: A list of a "str" and "Key" class instance.
+            threshold: Number of signatures needed to pass the verification.
+
+        Raises:
+            TypeError: If a key in "keys" is not an instance of the "Key" class.
+            ValueError: If "threshold" is not valid.
+            SignatureVerificationError: If the amount of "recognized_signers" is less
+                than provided threshold.
+
+        Returns:
+            recognized_signers: list of key names for which verification succeeds.
+        """
+
+        recognized_signers = []
+        used_keyids = []
+        pae = self.pae
+
+        # checks for threshold value.
+        if len(keys) < threshold or threshold <= 0:
+            raise ValueError(f"Threshold must be between 0 and {len(keys)}")
+
+        for signature in self.signatures:
+            for (name, key) in keys:
+                if not isinstance(key, Key):
+                    raise TypeError(f"expected type 'Key', got {type(key).__name__}")
+
+                # If key and signature include keyIDs but do not match skip.
+                if (
+                    signature.keyid is not None
+                    and key.keyid is not None
+                    and signature.keyid != key.keyid
+                ):
+                    continue
+
+                # If a key verifies the signature, we exit and use the result.
+                if key.verify(signature, pae):
+                    if key.keyid:
+                        # If keyid is already used, skip.
+                        if key.keyid in used_keyids:
+                            logger.debug(
+                                "Found envelope signed by different subkeys of the"
```

Comment:
This needn't always be because of subkeys, right? Am I missing something?

Reply:
[in-toto's verifylib](https://github.com/in-toto/in-toto/blob/f6a4f6239ef28fc482adf147af79f030086f63f0/in_toto/verifylib.py#L492-L496) do something similar, according to my findings it because of subkeys.

Reply:
Some thoughts:
1. That one is really only used in the instance of GPG keys, though it was generalized to all keys during the implementation. Because of certain in-toto nuances (as thresholds require multiple link files rather than multiple signatures on a single link file), it's going to be triggered in the GPG subkey scenario. We can't make those assumptions in the general case with DSSE. This is purely checking for multiple signatures from the same key, it doesn't mean that one signature was from the subkey of another.
2. As such I think the error message here should be more general--something to the effect of there's already one valid signature from the key, second signature will be skipped etc.
3. The error should definitely make no reference to "step threshold". That's very specific to in-toto and doesn't belong in the DSSE code.

Reply:
I have changed the message.

I don't know how this works, because I haven't looked at GPG code yet and it need to be implemented with DSSE.
I seen this case in [go-securesystemslib](https://github.com/secure-systems-lab/go-securesystemslib/blob/c1e41ba5f168bb0cecaac7ce1bbe787c20e72d9b/dsse/verify.go#L93-L97).

I think we should take a look again after implementation of `GPGSignature`.